affine/__init__.py

```diff
@@ -286,6 +286,62 @@ def determinant(self):
         return a * e - b * d
 
     @property
+    def _scaling(self):
+        """The absolute scaling factors of the transformation.
+
+        This tuple represents the absolute value of the scaling factors of the
+        transformation, sorted from bigger to smaller.
+        """
+        a, b, _, d, e, _, _, _, _ = self
+
+        # The singular values are the square root of the eigenvalues
+        # of the matrix times its transpose, M M*
+        # Computing trace and determinant of M M*
+        trace = a**2 + b**2 + d**2 + e**2
+        det = (a * e - b * d)**2
+
+        delta = trace**2 / 4 - det
+        if delta < 1e-12:
+            delta = 0
+
+        l1 = math.sqrt(trace / 2 + math.sqrt(delta))
+        l2 = math.sqrt(trace / 2 - math.sqrt(delta))
+        return l1, l2
+
+    @property
+    def eccentricity(self):
+        """The eccentricity of the affine transformation.
+
+        This value represents the eccentricity of an ellipse under
+        this affine transformation.
+
+        Raises NotImplementedError for improper transformations.
+        """
+        if self.is_proper or self.is_degenerate:
+            l1, l2 = self._scaling
+            return math.sqrt(l1 ** 2 - l2 ** 2) / l1
+        else:
+            raise NotImplementedError
```

Comment:
Rather than becoming undefined, it does not give the expected results.

```
In [13]: rx = Affine.scale(-1, 2)

In [14]: rx.determinant
Out[14]: -2.0

In [15]: rx._scaling
Out[15]: (2.0, 1.0)
```

However, depending on how you consider it, the eccentricity is right:

```
In [18]: class Bffine(Affine):
    ...:     @property
    ...:     def is_proper(self):
    ...:         return True
    ...:     

In [19]: rx = Bffine.scale(-1, 2)

In [20]: rx.determinant
Out[20]: -2.0

In [21]: rx.is_proper
Out[21]: True

In [22]: rx.eccentricity
Out[22]: 0.8660254037844386
```

so I might as well remove the check at all and allow the computation even for improper transformations. What do you think?